implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java

```diff
@@ -0,0 +1,261 @@
+package io.micrometer.registry.otlp;
+
+import io.micrometer.core.instrument.*;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.cumulative.*;
+import io.micrometer.core.instrument.distribution.*;
+import io.micrometer.core.instrument.distribution.pause.PauseDetector;
+import io.micrometer.core.instrument.internal.CumulativeHistogramLongTaskTimer;
+import io.micrometer.core.instrument.internal.DefaultGauge;
+import io.micrometer.core.instrument.internal.DefaultMeter;
+import io.micrometer.core.instrument.push.PushMeterRegistry;
+import io.micrometer.core.instrument.util.MeterPartition;
+import io.micrometer.core.instrument.util.NamedThreadFactory;
+import io.micrometer.core.ipc.http.HttpSender;
+import io.micrometer.core.ipc.http.HttpUrlConnectionSender;
+import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceRequest;
+import io.opentelemetry.proto.common.v1.AnyValue;
+import io.opentelemetry.proto.common.v1.KeyValue;
+import io.opentelemetry.proto.metrics.v1.*;
+import io.opentelemetry.proto.metrics.v1.Histogram;
+import io.opentelemetry.proto.resource.v1.Resource;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.function.ToDoubleFunction;
+import java.util.function.ToLongFunction;
+import java.util.stream.Collectors;
+
+/**
+ * Publishes meters in OTLP (OpenTelemetry Protocol) format.
+ *
+ * @since 1.9.0
+ */
+public class OtlpMeterRegistry extends PushMeterRegistry {
+
+    private static final ThreadFactory DEFAULT_THREAD_FACTORY = new NamedThreadFactory("otlp-metrics-publisher");
+    private final OtlpConfig config;
+    private final HttpSender httpSender;
+
+    private final Resource resource;
+
+    public OtlpMeterRegistry() {
+        this(OtlpConfig.DEFAULT, Clock.SYSTEM);
+    }
+
+    public OtlpMeterRegistry(OtlpConfig config, Clock clock) {
+        this(config, clock, new HttpUrlConnectionSender());
+    }
+
+    public OtlpMeterRegistry(OtlpConfig config, Clock clock, HttpSender httpSender) {
```

Comment:
It's come up before as users have wanted to write metrics to e.g. kafka before storing them.
There might not be an open issue for it. Would you mind making one so we can track it? I don't think it'll be done as part of this pull request, though.

I suppose if we plan to eventually support grpc with this registry, it may be a mistake to expose `HttpSender` in the public API now.

Reply:
That was my thought. If we don't have a good generic interface for `Sender` today, that's ok. But maybe we can encapsulate the usage of HttpSender into some protected method that is easily overridden for folks who have some other requirement.

Reply:
For now I've made the constructor private. We can consider what to expose as the use cases for customization are brought up by users. I think this gives us the most flexibility to evolve and add features while getting something usable out sooner.